Scylla sometimes fails to boot with node replace

In CI run https://jenkins.scylladb.com/job/scylla-master/job/build/1366/ the test `test/topology/test_topology_ip.py` failed once (it usually passes...) when Scylla failed to boot (in debug build):

```
E               Failed to start server with ID = 181, IP = 127.213.224.6.
E               Check the log files:
E               /jenkins/workspace/scylla-master/build/scylla/testlog/x86_64/test.py.debug-release-dev.log
E               /jenkins/workspace/scylla-master/build/scylla/testlog/x86_64/debug/scylla-181.log
```

The log file https://jenkins.scylladb.com/job/scylla-master/job/build/1366/artifact/testlog/x86_64/debug/scylla-181.log has some weird repair-related RPC error during the boot:

```
...
INFO  2023-02-15 04:46:38,853 [shard 0] storage_service - enable_repair_based_node_ops=true, allowed_repair_based_node_ops={replace}
INFO  2023-02-15 04:46:38,853 [shard 0] storage_service - replace[812e5015-9e25-42f0-9adc-8f9aca3bc3b7]: Using repair based node ops to sync data
INFO  2023-02-15 04:46:38,867 [shard 0] repair - replace_with_repair: started with keyspaces=system_authsystem_distributed_everywheresystem_distributedsystem_traces, source_dc=datacenter1, nr_ranges_total=438, ignore_nodes=[]
INFO  2023-02-15 04:46:38,869 [shard 0] repair - replace_with_repair: started with keyspace=system_auth, source_dc=datacenter1, nr_ranges=48, ignore_nodes=[]
INFO  2023-02-15 04:46:38,871 [shard 0] repair - replace_with_repair: finished with keyspace=system_auth, source_dc=datacenter1, nr_ranges=0
INFO  2023-02-15 04:46:38,871 [shard 0] repair - replace_with_repair: started with keyspace=system_distributed_everywhere, source_dc=datacenter1, nr_ranges=49, ignore_nodes=[]
INFO  2023-02-15 04:46:38,874 [shard 0] repair - repair[f43eeae1-6808-4ff0-abe6-f83c9c286292]: sync data for keyspace=system_distributed_everywhere, status=started
INFO  2023-02-15 04:46:38,875 [shard 0] repair - repair[f43eeae1-6808-4ff0-abe6-f83c9c286292]: Started to repair 1 out of 1 tables in keyspace=system_distributed_everywhere, table=cdc_generation_descriptions_v2, table_id=234d2227-dd63-3d37-ac5f-c013e2ea9e6e, repair_reason=replace
INFO  2023-02-15 04:46:38,876 [shard 1] repair - repair[f43eeae1-6808-4ff0-abe6-f83c9c286292]: Started to repair 1 out of 1 tables in keyspace=system_distributed_everywhere, table=cdc_generation_descriptions_v2, table_id=234d2227-dd63-3d37-ac5f-c013e2ea9e6e, repair_reason=replace
WARN  2023-02-15 04:46:38,892 [shard 1] repair - repair[f43eeae1-6808-4ff0-abe6-f83c9c286292]: shard=1, keyspace=system_distributed_everywhere, cf=cdc_generation_descriptions_v2, range=(-inf, -8681815950261682620], got error in row level repair: seastar::rpc::remote_verb_error (get_repair_meta: repair_meta_id 1 for node 127.213.224.6 does not exist)
WARN  2023-02-15 04:46:39,894 [shard 0] repair - repair[f43eeae1-6808-4ff0-abe6-f83c9c286292]: sync data for keyspace=system_distributed_everywhere, status=failed: std::runtime_error ({shard 0: seastar::rpc::remote_verb_error (get_repair_meta: repair_meta_id 0 for node 127.213.224.6 does not exist), shard 1: seastar::rpc::remote_verb_error (get_repair_meta: repair_meta_id 1 for node 127.213.224.6 does not exist)})
INFO  2023-02-15 04:46:39,989 [shard 0] storage_service - replace[812e5015-9e25-42f0-9adc-8f9aca3bc3b7]: Stopped heartbeat_updater
ERROR 2023-02-15 04:46:39,989 [shard 0] storage_service - replace[812e5015-9e25-42f0-9adc-8f9aca3bc3b7]: Abort replace operation started, replace_nodes={{127.213.224.6 -> 127.213.224.6}}, sync_nodes=[127.213.224.6, 127.213.224.54, 127.213.224.31], ignore_nodes=[]: std::runtime_error ({shard 0: seastar::rpc::remote_verb_error (get_repair_meta: repair_meta_id 0 for node 127.213.224.6 does not exist), shard 1: seastar::rpc::remote_verb_error (get_repair_meta: repair_meta_id 1 for node 127.213.224.6 does not exist)})
INFO  2023-02-15 04:46:39,993 [shard 0] storage_service - replace[812e5015-9e25-42f0-9adc-8f9aca3bc3b7]: Removed replacing_node=127.213.224.6 to replace existing_node=127.213.224.6, coordinator=127.213.224.6
INFO  2023-02-15 04:46:39,993 [shard 0] token_metadata - Removed node 127.213.224.6 as pending replacing endpoint which replaces existing node 127.213.224.6
ERROR 2023-02-15 04:46:40,016 [shard 0] storage_service - replace[812e5015-9e25-42f0-9adc-8f9aca3bc3b7]: Abort replace operation finished, replace_nodes={{127.213.224.6 -> 127.213.224.6}}, sync_nodes=[127.213.224.6, 127.213.224.54, 127.213.224.31], ignore_nodes=[]: std::runtime_error ({shard 0: seastar::rpc::remote_verb_error (get_repair_meta: repair_meta_id 0 for node 127.213.224.6 does not exist), shard 1: seastar::rpc::remote_verb_error (get_repair_meta: repair_meta_id 1 for node 127.213.224.6 does not exist)})
...
ERROR 2023-02-15 04:46:53,058 [shard 0] init - Startup failed: std::runtime_error ({shard 0: seastar::rpc::remote_verb_error (get_repair_meta: repair_meta_id 0 for node 127.213.224.6 does not exist), shard 1: seastar::rpc::remote_verb_error (get_repair_meta: repair_meta_id 1 for node 127.213.224.6 does not exist)})
```

@asias does this error look familiar? Should this problem really have caused Scylla to refuse to boot, or maybe it is some sort of race because the topology tests do not "sleep enough"? (obviously, they shouldn't need to sleep, but I remember we had bugs like this in the past).

@kostja @kbr-scylla @alecco maybe you can explain better than me what the failing test was trying to do when it failed?

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -2672,6 +2672,7 @@ public:
     sstring desc;
     locator::host_id host_id;   // Host ID of the node operand (i.e. added, replaced, or leaving node)
     inet_address endpoint;      // IP address of the node operand (i.e. added, replaced, or leaving node)
+    locator::token_metadata_ptr tmptr;
     std::unordered_set<gms::inet_address> sync_nodes;
     std::unordered_set<gms::inet_address> ignore_nodes;
     node_ops_cmd_request req;
@@ -2683,6 +2684,7 @@ public:
         : ss(ss_)
         , host_id(id)
         , endpoint(ep)
+        , tmptr(ss.get_token_metadata_ptr())
         , req(cmd, uuid)
         , heartbeat_interval(ss._db.local().get_config().nodeops_heartbeat_interval_seconds())
     {}
@@ -2697,8 +2699,25 @@ public:
         return req.ops_uuid;
     }
 
-    void start(sstring desc_) {
+    // may be called multiple times
+    void start(sstring desc_, std::function<bool(gms::inet_address)> sync_to_node = [] (gms::inet_address) { return true; }) {
         desc = std::move(desc_);
+
+        slogger.info("{}[{}]: Started {} operation: node={}/{}", desc, uuid(), desc, host_id, endpoint);
+
+        refresh_sync_nodes(std::move(sync_to_node));
+    }
+
+    void refresh_sync_nodes(std::function<bool(gms::inet_address)> sync_to_node = [] (gms::inet_address) { return true; }) {
+        // sync data with all normal token owners
+        sync_nodes.clear();
+        for (const auto& [node, host_id] : tmptr->get_endpoint_to_host_id_map_for_reading()) {
+            seastar::thread::maybe_yield();
+            if (!ignore_nodes.contains(node) && sync_to_node(node)) {
+                sync_nodes.insert(node);
+            }
+        }
+
         for (auto& node : sync_nodes) {
             if (!ss.gossiper().is_alive(node)) {
                 nodes_down.emplace(node);
@@ -2710,7 +2729,7 @@ public:
             throw std::runtime_error(msg);
         }
 
-        slogger.info("{}[{}]: Started {} operation: node={}/{}, sync_nodes={}, ignore_nodes={}", desc, uuid(), desc, host_id, endpoint, sync_nodes, ignore_nodes);
+        slogger.info("{}[{}]: sync_nodes={}, ignore_nodes={}", desc, uuid(), desc, host_id, endpoint, sync_nodes, ignore_nodes);
     }
 
     future<> stop() noexcept {
@@ -2898,9 +2917,14 @@ future<> storage_service::decommission() {
                 auto& db = ss._db.local();
                 node_ops_ctl ctl(ss, node_ops_cmd::decommission_prepare, db.get_config().host_id, ss.get_broadcast_address());
                 auto stop_ctl = deferred_stop(ctl);
-                auto tmptr = ss.get_token_metadata_ptr();
+
+                // Step 1: Decide who needs to sync data
+                // TODO: wire ignore_nodes provided by user
+                ctl.start("decommission");
+
                 uuid = ctl.uuid();
                 auto endpoint = ctl.endpoint;
+                const auto& tmptr = ctl.tmptr;
                 if (!tmptr->is_normal_token_owner(endpoint)) {
                     throw std::runtime_error("local node is not a member of the token ring yet");
                 }
@@ -2936,17 +2960,6 @@ future<> storage_service::decommission() {
 
                 slogger.info("DECOMMISSIONING: starts");
                 ctl.req.leaving_nodes = std::list<gms::inet_address>{endpoint};
-                // TODO: wire ignore_nodes provided by user
-
-                // Step 1: Decide who needs to sync data
-                for (const auto& [node, host_id] : tmptr->get_endpoint_to_host_id_map_for_reading()) {
-                    seastar::thread::maybe_yield();
-                    if (!ctl.ignore_nodes.contains(node)) {
-                        ctl.sync_nodes.insert(node);
-                    }
-                }
-
-                ctl.start("decommission");
 
                 assert(ss._group0);
                 raft_available = ss._group0->wait_for_raft().get();
@@ -3058,12 +3071,13 @@ void storage_service::run_bootstrap_ops(std::unordered_set<token>& bootstrap_tok
     node_ops_ctl ctl(*this, node_ops_cmd::bootstrap_prepare, db.get_config().host_id, get_broadcast_address());
     auto stop_ctl = deferred_stop(ctl);
     const auto& uuid = ctl.uuid();
+
+    // Step 1: Decide who needs to sync data for bootstrap operation
     // TODO: Specify ignore_nodes
+    ctl.start("bootstrap");
 
     auto start_time = std::chrono::steady_clock::now();
     for (;;) {
-        // Step 1: Decide who needs to sync data for bootstrap operation
-        ctl.sync_nodes = get_nodes_to_sync_with(ctl.ignore_nodes).get();
         ctl.sync_nodes.insert(get_broadcast_address());
 
         // Step 2: Wait until no pending node operations
@@ -3088,10 +3102,10 @@ void storage_service::run_bootstrap_ops(std::unordered_set<token>& bootstrap_tok
             }
             slogger.warn("bootstrap[{}]: Found pending node ops = {}, sleep 5 seconds and check again", uuid, pending_ops);
             sleep_abortable(std::chrono::seconds(5), _abort_source).get();
+            ctl.refresh_sync_nodes();
+            // the bootstrapping node will be added back when we loop
         }
     }
-
-    ctl.start("bootstrap");
 
     auto tokens = std::list<dht::token>(bootstrap_tokens.begin(), bootstrap_tokens.end());
     ctl.req.bootstrap_nodes = {
@@ -3121,14 +3135,14 @@ void storage_service::run_replace_ops(std::unordered_set<token>& bootstrap_token
     auto stop_ctl = deferred_stop(ctl);
     const auto& uuid = ctl.uuid();
     gms::inet_address replace_address = replace_info.address;
-    auto tmptr = get_token_metadata_ptr();
-    ctl.ignore_nodes = parse_node_list(_db.local().get_config().ignore_dead_nodes_for_replace(), *tmptr);
+    ctl.ignore_nodes = parse_node_list(_db.local().get_config().ignore_dead_nodes_for_replace(), *ctl.tmptr);
     // Step 1: Decide who needs to sync data for replace operation
-    ctl.sync_nodes = get_nodes_to_sync_with(ctl.ignore_nodes).get();
-    ctl.sync_nodes.erase(replace_address);
+    // The replacing node is not a normal token owner yet
+    // Add it back explicitly after checking all other nodes.
+    ctl.start("replace", [&] (gms::inet_address node) {
+        return node != replace_address;
+    });
     ctl.sync_nodes.insert(get_broadcast_address());
-
-    ctl.start("replace");
 
     auto sync_nodes_generations = _gossiper.get_generation_for_nodes(ctl.sync_nodes).get();
     // Map existing nodes to replacing nodes
@@ -3242,8 +3256,10 @@ future<> storage_service::removenode(locator::host_id host_id, std::list<locator
                 ss.raft_removenode(host_id).get();
                 return;
             }
-            auto uuid = node_ops_id::create_random_id();
-            auto tmptr = ss.get_token_metadata_ptr();
+            node_ops_ctl ctl(ss, node_ops_cmd::removenode_prepare, host_id, gms::inet_address());
+            auto stop_ctl = deferred_stop(ctl);
+            auto uuid = ctl.uuid();
+            const auto& tmptr = ctl.tmptr;
             auto endpoint_opt = tmptr->get_endpoint_for_host_id(host_id);
             assert(ss._group0);
             auto raft_id = raft::server_id{host_id.uuid()};
@@ -3270,17 +3286,15 @@ future<> storage_service::removenode(locator::host_id host_id, std::list<locator
                 throw std::runtime_error(message);
             }
 
+            for (auto& hoep : ignore_nodes_params) {
+                hoep.resolve(*tmptr);
+                ctl.ignore_nodes.insert(hoep.endpoint);
+            }
+
             bool removed_from_token_ring = !endpoint_opt;
             if (endpoint_opt) {
                 auto endpoint = *endpoint_opt;
-                node_ops_ctl ctl(ss, node_ops_cmd::removenode_prepare, host_id, endpoint, uuid);
-                auto stop_ctl = deferred_stop(ctl);
-                auto tokens = tmptr->get_tokens(endpoint);
-
-                for (auto& hoep : ignore_nodes_params) {
-                    hoep.resolve(*tmptr);
-                    ctl.ignore_nodes.insert(hoep.endpoint);
-                }
+                ctl.endpoint = endpoint;
 
                 // Step 1: Make the node a group 0 non-voter before removing it from the token ring.
                 //
@@ -3301,14 +3315,11 @@ future<> storage_service::removenode(locator::host_id host_id, std::list<locator
                 // If the user want the removenode opeartion to succeed even if some of the nodes
                 // are not available, the user has to explicitly pass a list of
                 // node that can be skipped for the operation.
-                for (const auto& [node, hostid] : tmptr->get_endpoint_to_host_id_map_for_reading()) {
-                    seastar::thread::maybe_yield();
-                    if (node != endpoint && !ctl.ignore_nodes.contains(node)) {
-                        ctl.sync_nodes.insert(node);
-                    }
-                }
+                ctl.start("removenode", [&] (gms::inet_address node) {
+                    return node != endpoint;
+                });
 
-                ctl.start("removenode");
+                auto tokens = tmptr->get_tokens(endpoint);
 
                 try {
                     // Step 3: Start heartbeat updater

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -1332,10 +1332,10 @@ future<> storage_service::mark_existing_views_as_built(sharded<db::system_distri
     });
 }
 
-std::unordered_set<gms::inet_address> storage_service::get_ignore_dead_nodes_for_replace(const token_metadata& tm) {
+std::unordered_set<gms::inet_address> storage_service::parse_node_list(sstring comma_separated_list, const token_metadata& tm) {
     std::vector<sstring> ignore_nodes_strs;
     std::unordered_set<gms::inet_address> ignore_nodes;
-    boost::split(ignore_nodes_strs, _db.local().get_config().ignore_dead_nodes_for_replace(), boost::is_any_of(","));
+    boost::split(ignore_nodes_strs, comma_separated_list, boost::is_any_of(","));
     for (std::string n : ignore_nodes_strs) {
         try {
             std::replace(n.begin(), n.end(), '\"', ' ');
@@ -1346,7 +1346,7 @@ std::unordered_set<gms::inet_address> storage_service::get_ignore_dead_nodes_for
                 ignore_nodes.insert(ep_and_id.endpoint);
             }
         } catch (...) {
-            throw std::runtime_error(format("Failed to parse --ignore-dead-nodes-for-replace parameter: ignore_nodes={}, node={}: {}", ignore_nodes_strs, n, std::current_exception()));
+            throw std::runtime_error(format("Failed to parse node list: {}: invalid node={}: {}", ignore_nodes_strs, n, std::current_exception()));
         }
     }
     return ignore_nodes;
@@ -1406,7 +1406,7 @@ future<> storage_service::bootstrap(cdc::generation_service& cdc_gen_service, st
 
         {
             // Wait for normal state handler to finish for existing nodes in the cluster.
-            auto ignore_nodes = replacement_info ? get_ignore_dead_nodes_for_replace(get_token_metadata())
+            auto ignore_nodes = replacement_info ? parse_node_list(_db.local().get_config().ignore_dead_nodes_for_replace(), get_token_metadata())
                                                  // TODO: specify ignore_nodes for bootstrap
                                                  : std::unordered_set<gms::inet_address>{};
             auto sync_nodes = get_nodes_to_sync_with(ignore_nodes).get();
@@ -3122,7 +3122,7 @@ void storage_service::run_replace_ops(std::unordered_set<token>& bootstrap_token
     const auto& uuid = ctl.uuid();
     gms::inet_address replace_address = replace_info.address;
     auto tmptr = get_token_metadata_ptr();
-    ctl.ignore_nodes = get_ignore_dead_nodes_for_replace(*tmptr);
+    ctl.ignore_nodes = parse_node_list(_db.local().get_config().ignore_dead_nodes_for_replace(), *tmptr);
     // Step 1: Decide who needs to sync data for replace operation
     ctl.sync_nodes = get_nodes_to_sync_with(ctl.ignore_nodes).get();
     ctl.sync_nodes.erase(replace_address);

--- a/service/storage_service.hh
+++ b/service/storage_service.hh
@@ -294,12 +294,13 @@ private:
     void run_replace_ops(std::unordered_set<token>& bootstrap_tokens, replacement_info replace_info);
     void run_bootstrap_ops(std::unordered_set<token>& bootstrap_tokens);
 
-    std::unordered_set<gms::inet_address> get_ignore_dead_nodes_for_replace(const locator::token_metadata& tm);
     future<std::unordered_set<gms::inet_address>> get_nodes_to_sync_with(
             const std::unordered_set<gms::inet_address>& ignore_dead_nodes);
     future<> wait_for_ring_to_settle(std::chrono::milliseconds delay);
 
 public:
+
+    static std::unordered_set<gms::inet_address> parse_node_list(sstring comma_separated_list, const locator::token_metadata& tm);
 
     future<> check_for_endpoint_collision(std::unordered_set<gms::inet_address> initial_contact_nodes,
             const std::unordered_map<gms::inet_address, sstring>& loaded_peer_features);

